source/posts/2019-06-25-GoCD-non-root-containers.html.markdown.erb

```diff
@@ -0,0 +1,182 @@
+  </figure>
+<% end %>
+
+
+Docker images run with root privileges by default. `root` user in the container is the same root (`uid:0`) as on the host machine. If a user manages to break out of an application running as `root` in a container, they may be able to gain access to the host machine with the same `root` user.
+
+Running containers using non-root user provides an additional layer of security against processes escaping the container due to a container engine vulnerability and thereby achieving escalated permissions on the host node.
+
+
+Starting from GoCD release `v19.6.0`, GoCD containers run as non-root user, by default. The Dockerized GoCD application will run as `go:root` (`uid:1000`, `gid:0`) user instead of running as `root:root` (`uid:0`, `gid:0`).
+
+## What are the changes made to GoCD Docker Images to support running containers as non-root?
+
+
+### <u>1. Add USER instruction in the Dockerfile:</u>
+
+The USER instruction sets the user name (or UID) to use when running the container and for any instructions that follow it in the Dockerfile.
+
+GoCD Dockerfile adds a USER instruction to change the current user from the `root` user to `go` (`uid:1000`) user. This instruction will make sure that the `go` (`uid:1000`) user is used while running the docker container and also for any `RUN`, `CMD` and `ENTRYPOINT` instructions that follow it in the Dockerfile.
+
+
+### <u>2. Add go user to root group:</u>
+
+OpenShift Container Platform (OCP) manages containers life-cycle with security in mind at every stage and every layer of application deployments.
+
+By default, OpenShift prevents root user containers by running containers using an arbitrarily assigned user ID. This arbitrary user ID in the container is always a member of `root` group on OCP. For an image to support running as an arbitrary user, directories and files that may be written to by processes in the image should be owned by the root group and be read/writable by that group.
+
+Hence, to allow GoCD application to be deployed on OCP, dockerized GoCD applications run with `go:root` user.
+
+**Note** : The root group does not have any special permissions (unlike the `root` user). So, there are no security concerns with this arrangement.
+
+For more information, checkout [OCP General Container Image Guidelines](https://docs.openshift.com/container-platform/3.5/creating_images/guidelines.html).
+
+
+### <u>3. Modify GoCD Directories and File permissions:</u>
+
+The GoCD server stores all configuration, pipeline history database, artifacts, plugins, addons and logs into `/godata`, whereas, additional information such as secure credentials, ssh keys can additionally be stored at `/home/go`.
+
+As the existing container user has changed from `go:go` to `go:root`, the file permissions of the above mentioned directories and files are changed to `go:root`.
```

Comment:
Should we mention about the case of NFS mount path if mapped to `/godata` server will fail to start up since the mount path permissions by default will be `root:root`? User will have to update it to `go:root` to get the server running. Not sure if we should mention it here or as breaking change in release notes

Reply:
@rajiesh -- Hmm. It's better to document than to assume no one will face this issue.
Will do it :)